File: bindings/octave/lib/Logger.m
classdef Logger
    %Logger  Thin, stateless view over the library's global logger (Octave).
    %
    %   Mirrors Python's ptiff.Logger and Ruby's PTiff::Logger: a small
    %   object whose methods call the SWIG low-level ptiff_logger_* functions.
    %   The underlying C logger is a single process-wide instance, so any
    %   number of Logger objects read and mutate it.
    %
    %   Level numeric values are the PTIFF_LOG_* C ABI enum values, exposed as
    %   read-only properties (TRACE..OFF). Convenience methods trace/debug/...
    %   emit at those levels.
    %
    %   Example
    %   -------
    %       l = ptiff.Logger();
    %       l.set_level(l.ERROR);
    %       l.error("boom");
    %
    %   Requires the SWIG module (call ptiff() once first).

    properties (Constant = true)
        % PTIFF_LOG_* level values (match the C ABI enum).
        TRACE    = 0
        DEBUG    = 1
        INFO     = 2
        WARN     = 3
        ERROR    = 4
        CRITICAL = 5
        OFF      = 6
    end

    methods
        function level = get_level(obj)
            %GET_LEVEL  Current minimum level that gets emitted.
            level = ptiff_logger_level();
        end

        function l = level(obj)
            %LEVEL  Alias for get_level.
            l = ptiff_logger_level();
        end

        function set_level(obj, value)
            %SET_LEVEL  Set the minimum level that gets emitted.
            ptiff_logger_set_level(int32(value));
        end

        function log(obj, value, msg)
            %LOG  Emit msg at level value (filtered below the threshold).
            ptiff_logger_log(int32(value), char(msg));
        end

        function trace(obj, msg),    obj.log(obj.TRACE, msg);  end
        function debug(obj, msg),    obj.log(obj.DEBUG, msg);  end
        function info(obj, msg),     obj.log(obj.INFO, msg);   end
        function warning(obj, msg),  obj.log(obj.WARN, msg);   end
        function error(obj, msg),    obj.log(obj.ERROR, msg);  end
        function critical(obj, msg), obj.log(obj.CRITICAL, msg); end
    end
end

File: bindings/python/src/ptiff/logger.py
"""Logging wrapper over the SWIG low-level logger surface.

The C ABI exposes three free functions (`ptiff_logger_set_level`,
`ptiff_logger_level`, `ptiff_logger_log`) and the `PTIFF_LOG_*` level
constants. This module wraps them as a tiny ``Logger`` value object so the
idiomatic object layer stays uniform with `Camera`/`Image`/`Tile`.
"""

from .ptiff import (
    PTIFF_LOG_CRITICAL,
    PTIFF_LOG_DEBUG,
    PTIFF_LOG_ERROR,
    PTIFF_LOG_INFO,
    PTIFF_LOG_OFF,
    PTIFF_LOG_TRACE,
    PTIFF_LOG_WARN,
    ptiff_logger_level,
    ptiff_logger_log,
    ptiff_logger_set_level,
)

TRACE = PTIFF_LOG_TRACE
DEBUG = PTIFF_LOG_DEBUG
INFO = PTIFF_LOG_INFO
WARN = PTIFF_LOG_WARN
ERROR = PTIFF_LOG_ERROR
CRITICAL = PTIFF_LOG_CRITICAL
OFF = PTIFF_LOG_OFF


class Logger:
    """A thin, stateless view over the library's global logger.

    The underlying C logger is a single process-wide instance, so these
    methods read and mutate that one instance regardless of how many
    ``Logger`` objects exist.
    """

    # Level constants exposed on the class for convenience.
    TRACE = TRACE
    DEBUG = DEBUG
    INFO = INFO
    WARN = WARN
    ERROR = ERROR
    CRITICAL = CRITICAL
    OFF = OFF

    @property
    def level(self) -> int:
        return ptiff_logger_level()

    @level.setter
    def level(self, value: int) -> None:
        ptiff_logger_set_level(int(value))

    def set_level(self, level: int) -> None:
        """Set the minimum level that gets emitted (see the PTIFF_LOG_* values)."""
        ptiff_logger_set_level(int(level))

    def log(self, level: int, message: str) -> None:
        """Emit ``message`` at ``level``; filtered out below the current threshold."""
        ptiff_logger_log(int(level), message)

    def trace(self, message: str) -> None:
        self.log(self.TRACE, message)

    def debug(self, message: str) -> None:
        self.log(self.DEBUG, message)

    def info(self, message: str) -> None:
        self.log(self.INFO, message)

    def warning(self, message: str) -> None:
        self.log(self.WARN, message)

    def error(self, message: str) -> None:
        self.log(self.ERROR, message)

    def critical(self, message: str) -> None:
        self.log(self.CRITICAL, message)
